itsm_api/views.py

```diff
@@ -806,6 +791,120 @@ def _post(self, data):
+        #
+        # <attachment_id>__<filename>
+        #
+        # For example:
+        # 1__errors.log
+        #
+        # The directory structure of saved attachment files looks like this:
+        #
+        # attachment_storage/
+        #     ticket__<ticket_id>/
+        #         <unique_filename>
+        #
+        # For example:
+        #
+        # attachment_storage/
+        #     ticket__1/
+        #         1__errors.log
+        #
+        #     ticket__2/
+        #         2__some_image.jpg
+        #         3__some_doc.docx
+        #
+        #     ticket__3/
+        #         4__requirements.txt
+        #
+        # Note here that the numbered directories are Ticket IDs, and the numbers prefixing
+        # the file name are the attachment IDs.
+
+        # We first need to get the base64 encoded attachment out of the data, so that the
+        # attachment file is not saved to disk twice (as a file and as a base64 string)
+        attachment_data   = data.pop('attachment_data')
+        new_attachment_id = DB_ATTACHMENT_TABLE.insert(data)
+
+        try:
+            # Decode the attachment data
+            decoded_attachment_data = base64.b64decode(attachment_data)
+
+            # The filename is expected to exist in the data because it is a mandatory key.
+            filename        = data['filename']
+            unique_filename = f"{new_attachment_id}__{filename}"
+
+            # Create the paths to the directory and the attachment file
+            dir_path     = os.path.join(_ATTACHMENT_FOLDER,
+                                        f"ticket__{str(data['ticket_id'])}")
+            path_to_file = os.path.join(dir_path, unique_filename)
+
+            # Make the directory for the attachment if it doesn't already exist, then save the
+            # attachment file.
+            os.makedirs(dir_path, exist_ok=True)
+            with open(path_to_file, "wb") as attachment_file:
+                attachment_file.write(decoded_attachment_data)
+        except OSError:
+            # Trying to save the decoded attachment file to disk went wrong. Rollback the
+            # database entry and return a 500 error.
+            # Note that since Python 3.3, IOError became an alias for OSError, hence OSError is
+            # the actual exception that will occur.
+            DB_ATTACHMENT_TABLE.remove(doc_ids=[new_attachment_id])
+            flask_restful.abort(500, message="Error occured while trying to save attachment "
+                                             "file data")
+        except Exception:
+            # Some error occured while trying to decode the attachment file data. Rollback the
+            # database entry and return a 500 error.
+            DB_ATTACHMENT_TABLE.remove(doc_ids=[new_attachment_id])
+            flask_restful.abort(500, message="Error occured while trying to decode "
+                                             "attachment file data")
+        else:
```

Comment:
Why do we even need the else branch here? If I understand correctly, `flash_restful.abort()` will end the execution, will it not? Alternatively, we can also remove else and just move line 904 and 905 to be in line 889.

Reply:
Ah right, I was thinking along the lines of "we only want to return the ID if it works", but I can see that if we're catching `Exception`, then `else` doesn't add any extra value. Updated to remove the else block